viam-orb-slam3/orbslam_server_v1.cc

```diff
@@ -146,6 +146,111 @@ ::grpc::Status SLAMServiceImpl::GetInternalStateStream(
+    q = extra->mutable_fields()->operator[]("quat").mutable_struct_value();
+    q->mutable_fields()->operator[]("real").set_number_value(actualPose[3]);
+    q->mutable_fields()->operator[]("imag").set_number_value(actualPose[0]);
+    q->mutable_fields()->operator[]("jmag").set_number_value(actualPose[1]);
+    q->mutable_fields()->operator[]("kmag").set_number_value(actualPose[2]);
+
+    response->set_component_reference(camera_name);
+
+    return grpc::Status::OK;
+}
+
+::grpc::Status SLAMServiceImpl::GetPointCloudMap(
+    ServerContext *context, const GetPointCloudMapRequest *request,
+    ServerWriter<GetPointCloudMapResponse> *writer) {
+    std::vector<ORB_SLAM3::MapPoint *> actualMap;
+    {
+        std::lock_guard<std::mutex> lk(slam_mutex);
+        actualMap = currMapPoints;
+    }
+
+    if (actualMap.size() == 0) {
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                            "currently no map points exist");
+    }
+
+    auto buffer = utils::PcdHeader(actualMap.size());
+
+    for (auto p : actualMap) {
+        Eigen::Matrix<float, 3, 1> v = p->GetWorldPos();
+        utils::WriteFloatToBufferInBytes(buffer, v.x());
+        utils::WriteFloatToBufferInBytes(buffer, v.y());
+        utils::WriteFloatToBufferInBytes(buffer, v.z());
+    }
+
+    std::string pcd_chunk;
+    GetPointCloudMapResponse response;
+    for (int start_index = 0; start_index < buffer.size();
+         start_index += maximumGRPCByteChunkSize) {
+        pcd_chunk = buffer.substr(start_index, maximumGRPCByteChunkSize);
+        response.set_point_cloud_pcd_chunk(pcd_chunk);
+        bool ok = writer->Write(response);
+        if (!ok)
+            return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                                "error while writing to stream: stream closed");
+    }
+    return grpc::Status::OK;
+}
+
+::grpc::Status SLAMServiceImpl::GetInternalState(
+    ServerContext *context, const GetInternalStateRequest *request,
+    ServerWriter<GetInternalStateResponse> *writer) {
+    std::stringbuf buffer;
+    // deferring reading the osa file in chunks until we run into issues
+    // with loading the file into memory
+    bool success = ArchiveSlam(buffer);
+    if (!success)
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                            "SLAM is not yet initialized");
+
+    std::string internal_state_chunk;
+    GetInternalStateResponse response;
+    std::string buffer_str = buffer.str();
+    for (int start_index = 0; start_index < buffer_str.size();
+         start_index += maximumGRPCByteChunkSize) {
+        internal_state_chunk =
+            buffer_str.substr(start_index, maximumGRPCByteChunkSize);
+        response.set_internal_state_chunk(internal_state_chunk);
+        bool ok = writer->Write(response);
+        if (!ok)
+            return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                                "error while writing to stream: stream closed");
```

Comment:
[q] Is there any situation where we would have trouble writing to the stream but it is not closed?

Reply:
Not that I am aware of, but I haven’t looked to deep into the writer